L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc

```diff
@@ -153,9 +208,167 @@ void L1NNTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descripti
+  if (tmpmass1 < 0)
+    tmpmass1 = 0;
+  L1TauEmu::pt_t mass = l1ct::pt_t(tmpmass1);
+  if (pt < fSeedPt_)
+    return;
+  result_t NN = fTauNNIdHW_->compute(seed, parts);
+  input_t* lNNVector = fTauNNIdHW_->NNVectorVar();
+  float pNNVec[80];
+  for (unsigned i0 = 0; i0 < 80; i0++)
+    pNNVec[i0] = float(lNNVector[i0]);
+  //std::cout << "HW NN " << lNNVector[0] << " : " <<  lNNVector[1] << " : " <<  lNNVector[2] << std::endl;
+  //std::cout << "HW NN " << pNNVec   [0] << " : " <<  pNNVec   [1] << " : " <<  pNNVec   [2] << std::endl;
+  L1TauEmu::etaphi_t eta = etaphi_t(seed.eta() * L1TauEmu::etaphi_base);
+  L1TauEmu::etaphi_t phi = etaphi_t(seed.phi() * L1TauEmu::etaphi_base);
+  math::PtEtaPhiMLorentzVector tempP4(
+      float(pt), float(eta) / L1TauEmu::etaphi_base, float(phi) / L1TauEmu::etaphi_base, float(mass));
+  l1t::PFTau l1PFTau(tempP4, pNNVec, NN, 0, lId);
+  l1PFTau.setZ0(float(z0) * 0.05);    //L1TauEmu::z0_base);
+  l1PFTau.setDxy(float(dxy) * 0.05);  //L1TauEmu::dxy_base);
+
+  uint32_t lPt = pt.to_uint();
+  uint32_t lEta = eta.to_uint();
+  uint32_t lPhi = phi.to_uint();
+  uint32_t lNN = NN.to_uint();
+  //printTau(lPt,lEta,lPhi,lNN);
+  iTaus->push_back(l1PFTau);
+}
+void L1NNTauProducer::printParts(std::vector<l1t::PFCandidate>& parts) {
+  for (unsigned i0 = 0; i0 < parts.size(); i0++) {
+    fprintf(file1_, " %16lx", parts[i0].encodedPuppi64());
+    if (i0 % 36 == 35)
+      fprintf(file1_, "\n");
+    if (i0 == 6 * 36 - 1)
+      break;
+    //std::cout << "--> test " << parts[i0].encodedPuppi64();
+    //if(i0 % 36 == 35)   std::cout << std::endl;
+  }
+  for (unsigned i0 = parts.size(); i0 < 6 * 36 - 1; i0++) {
+    unsigned long int dummy = 0;
+    fprintf(file1_, " %16lx", dummy);
+  }
+}
+void L1NNTauProducer::printTau(uint32_t& iPt, uint32_t& iEta, uint32_t& iPhi, uint32_t& iNN) {
+  fprintf(file2_, " %08x", iPt);
+  fprintf(file2_, " %08x", iEta);
+  fprintf(file2_, " %08x", iPhi);
+  fprintf(file2_, " %08x", iNN);
+  fprintf(file2_, "\n");
+}
+
+void L1NNTauProducer::process_HW(const l1t::PFCandidateCollection& parts,
+                                 std::unique_ptr<l1t::PFTauCollection>& iTaus) {
+  // The fixed point algorithm emulation
+  using namespace L1TauEmu;
+  std::vector<l1t::PFCandidate> work;
+  work.resize(parts.size());
+  std::transform(parts.begin(), parts.end(), work.begin(), [](const l1t::PFCandidate& part) { return part; });
+  std::sort(work.begin(), work.end(), [](l1t::PFCandidate i, l1t::PFCandidate j) {
+    return (l1ct::pt_t(i.pt()) > l1ct::pt_t(j.pt()));
+  });
+  //printParts(work);
+  std::vector<l1t::PFCandidate> seeds;
+  uint lSeed = l1t::PFCandidate::ChargedHadron;
+  if (fEMSeed)
+    lSeed = l1t::PFCandidate::Photon;
+  std::copy_if(work.begin(), work.end(), std::back_inserter(seeds), [&](const l1t::PFCandidate& part) {
+    return ((part.id() == l1t::PFCandidate::Electron || part.id() == l1t::PFCandidate::ChargedHadron ||
+             part.id() == lSeed) &&
```

Comment:
This logic looks ok, but it's a bit confusing so want to make sure there isn't a bug.  When `fEMSeed` is `false`, this line becomes `(part.id() == l1t::PFCandidate::Electron || part.id() == l1t::PFCandidate::ChargedHadron || part.id() == [lSeed](l1t::PFCandidate::ChargedHadron)` i.e. the last conditional is repeated.  Is that intended?